loki/transformations/block_index_transformations.py

```diff
@@ -414,3 +418,264 @@ def process_kernel(self, routine, targets, exclude_arrays):
+                loop_to_lower = Transformer({call: None for call in calls_within_loop}).visit(loop_to_lower)
+                # symbols that are defined or rather assigned within the loop
+                defined_symbols_loop = [assign.lhs for assign in FindNodes(ir.Assignment).visit(loop_to_lower.body)]
+                call_routine_variables = [var.name.lower() for var in FindVariables().visit(call.routine.body)]
+                additional_vars = [var for var in FindVariables().visit(loop_to_lower)
+                        if var.name.lower() != loop.variable.name.lower()
+                        and var not in call_arg_map and var.name not in call.routine.variable_map
+                        and var.name.lower() not in call_routine_variables
+                        # and var not in call_arg_map
+                        and isinstance(var, sym.Scalar) and var not in defined_symbols_loop]
+                additional_kwargs = as_tuple([(var.name, var.clone()) for var in additional_vars])
+                calls_within_loop = [call for call in  FindNodes(ir.CallStatement).visit(loop_to_lower.body)
+                        if str(call.name).lower() in targets]
+                if call.routine.name not in processed_routines:
+                    call.routine.body = ir.Section(as_tuple(Transformer({calls_within_loop[0]:\
+                            call.routine.body}).visit(loop_to_lower)))
+                    call.routine.arguments += as_tuple(var.clone(scope=call.routine,
+                        type=var.type.clone(intent='in')) for var in additional_vars)
+                call_arg_iter = list(call.arg_iter())
+                ignore_symbols = [loop.variable.name.lower()] +\
+                        [symbol.name.lower() for symbol in defined_symbols_loop]
+                _arguments = tuple(arg for i_arg, arg in enumerate(call.arguments)
+                        if call_arg_iter[i_arg][0].name.lower() not in ignore_symbols)
+                _kwarguments = tuple(kwarg for kwarg in call.kwarguments if kwarg[0].lower() not in ignore_symbols)
+                call._update(arguments=_arguments, kwarguments=_kwarguments+additional_kwargs)
+                # add a loki pragma to keep loop information in driver
+                call_pragmas = (self.generate_pragma(loop),)
+                call._update(pragma=(call.pragma if call.pragma else ()) + call_pragmas)
+                processed_routines += (call.routine.name,)
+                to_local_var[call.routine.name] = defined_symbols_loop + [loop.variable]
+            if lower_loop:
+                driver_loop_map[loop] = loop.body
+        routine.body = Transformer(driver_loop_map).visit(routine.body)
+        for call in calls:
+            if str(call.name).lower() not in targets:
+                continue
+            for var in to_local_var[call.routine.name]:
+                self.local_var(call, var)
+        # TODO: remove
+        self.remove_openmp_pragmas(routine)
```

Comment:
This should likely happen elsewhere in the processing pipeline

Reply:
True! Would you be ok with me changing this in the next PR?

Reply:
It adds a very unexpected behaviour to a transformation that appears otherwise to be more generally applicable. If it's not too inconvenient I would prefer this to be removed, or at least to limit the processing to the transformed loop? (E.g., use `with_pragmas_attached(ir.Loop)` and then strip the pragma only from the processed block loop)